File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def maxDiff(root: TreeNode, maxNode: int, minNode: int) -> int:
            if not root:
                return maxNode - minNode
            minNode, maxNode = min(minNode, root.val), max(maxNode, root.val)
            return max(maxDiff(root.left, maxNode, minNode), maxDiff(root.right, maxNode, minNode))
        return maxDiff(root, 0, 1e9)

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxDiff(TreeNode root, int maxNode, int minNode) {
        if(root == null) return maxNode - minNode;
        minNode = Math.min(minNode, root.val);
        maxNode = Math.max(maxNode, root.val);
        return Math.max(maxDiff(root.left, maxNode, minNode), maxDiff(root.right, maxNode, minNode));
    }
    public int maxAncestorDiff(TreeNode root) {
        return maxDiff(root, 0, 100000);
    }
}